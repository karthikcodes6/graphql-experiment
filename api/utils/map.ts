/**
 * Helper functions that order data records so that they
 * match the provided keys array.
 *
 * @see https://github.com/graphql/dataloader
 * @copyright 2016-present Kriasoft (https://git.io/Jt7GM)
 */

export function mapTo<R, K>(
  records: ReadonlyArray<R>,
  keys: ReadonlyArray<K>,
  keyFn: (record: R) => K,
): Array<R | null> {
  const map = new Map(records.map((x) => [keyFn(x), x]));
  return keys.map((key) => map.get(key) || null);
}

export function mapToMany<R, K>(
  records: ReadonlyArray<R>,
  keys: ReadonlyArray<K>,
  keyFn: (record: R) => K,
): Array<R[]> {
  const group = new Map<K, R[]>(keys.map((key) => [key, []]));
  records.forEach((record) => (group.get(keyFn(record)) || []).push(record));
  return Array.from(group.values());
}

export function mapToValues<R, K, V>(
  records: ReadonlyArray<R>,
  keys: ReadonlyArray<K>,
  keyFn: (record: R) => K,
  valueFn: (record?: R) => V,
): Array<V> {
  const map = new Map(records.map((x) => [keyFn(x), x]));
  return keys.map((key) => valueFn(map.get(key)));
}
