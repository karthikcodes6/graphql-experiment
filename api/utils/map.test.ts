/**
 * @copyright 2016-present Kriasoft (https://git.io/Jt7GM)
 */

import { mapTo, mapToMany, mapToValues } from "./map";

test("mapTo()", () => {
  const result = mapTo(
    [
      { id: 2, name: "b" },
      { id: 1, name: "a" },
    ],
    [1, 2],
    (x) => x.id,
  );
  expect(result).toMatchSnapshot();
});

test("mapToMany()", () => {
  const result = mapToMany(
    [
      { id: 2, name: "b" },
      { id: 1, name: "a" },
      { id: 1, name: "c" },
    ],
    [1, 2],
    (x) => x.id,
  );
  expect(result).toMatchSnapshot();
});

test("mapToValues()", () => {
  const result = mapToValues(
    [
      { id: 2, name: "b" },
      { id: 1, name: "a" },
      { id: 3, name: "c" },
    ],
    [1, 2, 3, 4],
    (x) => x.id,
    (x) => x?.name || null,
  );
  expect(result).toMatchSnapshot();
});
