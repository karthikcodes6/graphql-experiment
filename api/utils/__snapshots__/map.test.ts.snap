// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapTo() 1`] = `
Array [
  Object {
    "id": 1,
    "name": "a",
  },
  Object {
    "id": 2,
    "name": "b",
  },
]
`;

exports[`mapToMany() 1`] = `
Array [
  Array [
    Object {
      "id": 1,
      "name": "a",
    },
    Object {
      "id": 1,
      "name": "c",
    },
  ],
  Array [
    Object {
      "id": 2,
      "name": "b",
    },
  ],
]
`;

exports[`mapToValues() 1`] = `
Array [
  "a",
  "b",
  "c",
  null,
]
`;
